### PR TITLE
[PJRT:GPU] Treat GPU collective memory space as device memory space

### PR DESCRIPTION
**Pull request by zhenying-liu**
This is a regression fix when using --xla_gpu_enable_nccl_user_buffers=true.
Return device memory space when collective memory space is used as an output on GPU.

### CODE REVIEW

**Comment by NaiyerRizz** [comment]
Hi @zhenying-liu 
Seems like there are some conflicts in the branch, Can you please resolve those.
Thank you

**Comment by zhenying-liu** [comment]
> Hi @zhenying-liu Seems like there are some conflicts in the branch, Can you please resolve those. Thank you

Thanks. Resolved conflicts.

**Review by hawkinsp — APPROVED** [review]

**Comment by cheshire** [comment]
Any chance we could separate out XLA and PJRT changes? 

I'm somewhat new to that code, but hardcoding memory spaces in layout.h seems strange, is that the usual approach?

**Comment by cheshire** [comment]
Is it possible to describe in general terms, not in specific fixes, what the change is doing? Is it fixing a crash? Is it performance improvement? If it's a bullet list of five things, maybe it should be 5 PRs?

**Comment by zhenying-liu** [comment]
> Any chance we could separate out XLA and PJRT changes?
> 
> I'm somewhat new to that code, but hardcoding memory spaces in layout.h seems strange, is that the usual approach?

It's defined in XLA. It's harder to separate XLA and PJRT. But I can try my best to split.
I think defining memory spaces in layout.h is the best way that I can think of. Otherwise, defining it in different files could cause the conflicts. Currently both fast memory space and collect memory space has the value 1.

**Comment by zhenying-liu** [comment]
> Is it possible to describe in general terms, not in specific fixes, what the change is doing? Is it fixing a crash? Is it performance improvement? If it's a bullet list of five things, maybe it should be 5 PRs?

There is a crash when using NCCL user buffer: --xla_gpu_enable_nccl_user_buffers=true. The error message is: "INVALID_ARGUMENT: Unexpected memory space 1 in output layout". It happened in pjrt_stream_executor_client.cc:MemoryKindFromSimpleShape().  
Jax doesn't support any memory kind other than device and pinned_host. In this case, the output memory kind is neither device nor pinned_host. Our strategy is to treat CollectiveMemorySpace as device memory space to please Jax. But inside XLA, we keep using collective memory space to call nccl memory allocator to handle allocation/deallocation. The existing memory space color for collective memory space is 1, conflicting with Layout::kGenericFastMemorySpace. So we define a constant kCollectiveMemorySpace in layout.h with a unique value 6, and use this constant whenever it is needed.

**Comment by nouiz** [comment]
> > Is it possible to describe in general terms, not in specific fixes, what the change is doing? Is it fixing a crash? Is it performance improvement? If it's a bullet list of five things, maybe it should be 5 PRs?
> 
> There is a crash when using NCCL user buffer: --xla_gpu_enable_nccl_user_buffers=true. The error message is: "INVALID_ARGUMENT: Unexpected memory space 1 in output layout". It happened in pjrt_stream_executor_client.cc:MemoryKindFromSimpleShape(). Jax doesn't support any memory kind other than device and pinned_host. In this case, the output memory kind is neither device nor pinned_host. Our strategy is to treat CollectiveMemorySpace as device memory space to please Jax. But inside XLA, we keep using collective memory space to call nccl memory allocator to handle allocation/deallocation. The existing memory space color for collective memory space is 1, conflicting with Layout::kGenericFastMemorySpace. So we define a constant kCollectiveMemorySpace in layout.h with a unique value 6, and use this constant whenever it is needed.

This is a regression fix when using --xla_gpu_enable_nccl_user_buffers=true
Jane, can you update the description to tell that?

**Comment by nouiz** [comment]
> Is it possible to describe in general terms, not in specific fixes, what the change is doing? Is it fixing a crash? Is it performance improvement? If it's a bullet list of five things, maybe it should be 5 PRs?

This is a regression fix. Google JAX team want to make a releases, so merging this soon would be great.
@hawkinsp approved the changes, so all PJRT part should be good.
Splitting into multiple from outside would take extra times and this compete with doing a release of JAX very soon. So I let you @cheshire and @hawkinsp discuss that. Having regression in public version isn't great.

File: xla/layout.h
Comment (cheshire) [inline xla/layout.h:407]: Does this enum have to live in `layout.h` ? Maybe we could move it to a GPU-specific file? After all , GPU has to not collide only with GPU. Also having to recompiling XLA on any changes to this enum does not sound great.

**Review by cheshire — CHANGES_REQUESTED** [review]

File: xla/layout.h
Comment (jaro-sevcik) [reply xla/layout.h:407]: At the same time, it would be good if the constants would not collide, so I would prefer to have all the constants in one place. How about creating a new header file just for the memory space constants?

Comment (jaro-sevcik) [reply xla/layout.h:407]: @cheshire Conceptually what you suggest sounds consistent. I would be concerned that someone takes a layout from an array, tries to use for some other device and gets surprising results without any warning, but perhaps that's already not possible because one cannot really sneak in memory spaces in layout.

In practice, this PR is holding up a JAX release, so we should decide what we want soon. Are you proposing to undo the move/rename of the kCollectiveMemorySpaceColor constant and change of its value to 6 (without worrying about the collective space id 1 colliding with kGenericFastMemorySpace) and just keep the changes to xla/pjrt/pjrt_stream_executor_client.cc and to xla/pjrt/gpu/se_gpu_pjrt_client_test.cc?

File: xla/pjrt/gpu/se_gpu_pjrt_client_test.cc
Comment (cheshire) [inline xla/pjrt/gpu/se_gpu_pjrt_client_test.cc:1253]: What happens if we execute it without the changes in this PR?

Comment (zhenying-liu) [reply xla/pjrt/gpu/se_gpu_pjrt_client_test.cc:1253]: It will have the assertion failure in pjrt_stream_execution_client.cc:MemoryKindFromSimpleShape():
`INVALID_ARGUMENT: Unexpected memory space 1 in output layout`

File: xla/layout.h
Comment (zhenying-liu) [reply xla/layout.h:407]: @cheshire I have the smallest changes to xla/pjrt/pjrt_stream_executor_client.cc and xla/pjrt/gpu/se_gpu_pjrt_client_test.cc ready. With that one, I have to use kGenericFastMemorySpace for collective space id 1. If so, we can fix the regression now and do the clean up the definition later.

Comment (zhenying-liu) [reply xla/layout.h]: @cheshire Please take another look at the code review as I changed the code as mentioned above.

**Comment by nouiz** [comment]
Note, this PR was updated with the minimal fix to the regression.
The refactoring/code clean-up can be done after the JAX releases.
@cheshire @hawkinsp

**Review by cheshire — APPROVED** [review]

**Comment by nouiz** [comment]
This approved PR is failing in Google internal import.
Who can look at it? @xla-rotation

File: xla/pjrt/gpu/se_gpu_pjrt_client_test.cc
Comment (akuegel) [inline xla/pjrt/gpu/se_gpu_pjrt_client_test.cc:1239]: void function 'TestBody' should not return a value [-Wreturn-mismatch]
 1239 |   TF_ASSIGN_OR_RETURN(

**Review by akuegel — COMMENTED** [review]
Merging this PR is blocked on a compile error. I wonder whether this actually compiles on your end?

File: xla/pjrt/gpu/se_gpu_pjrt_client_test.cc
Comment (zhenying-liu) [reply xla/pjrt/gpu/se_gpu_pjrt_client_test.cc]: Thanks for pointing this out. Sorry for the mistake. Fixed.

**Review by akuegel — APPROVED** [review]

**Pull request closed without merging**
